File: android/app/src/main/java/io/github/atlashealthapp/atlas/ble/AlarmPlan.java
package io.github.atlashealthapp.atlas.ble;

import android.content.SharedPreferences;

import java.util.Calendar;

/**
 * Several alarms, and which one tonight belongs to.
 *
 * <p>Atlas held exactly one alarm until 2026-08-24. The ask was a weekday alarm
 * and a different weekend one.
 *
 * <p><b>The band really does have slots</b> - proven on the device, not assumed:
 * a write to slot 1 was accepted, the strap buzzed for it, and Zepp showed both
 * alarms afterwards. So each alarm owns a slot and the band rings each one
 * itself. Nothing is rewritten as the week turns.
 *
 * <h2>Why this resolves into the old keys instead of replacing them</h2>
 *
 * <p>{@link SmartAlarm} reads one alarm out of {@code KEY_HOUR} and friends, and
 * every piece of overnight state hangs off that - the window, the fired-once
 * mark, the re-arm debt. Rewriting all of it to be per-alarm would put the
 * riskiest code in the app through a change nobody can test except by sleeping.
 *
 * <p>So this picks tonight's alarm out of the list and writes it into exactly
 * the keys that already existed. {@code SmartAlarm} is untouched and still
 * reasons about one alarm; only the choosing is new. The one addition is
 * {@link #KEY_SLOT}, so a write lands on that alarm's own slot rather than
 * always on slot 0.
 *
 * <p><b>That is only sound because at most one alarm may be due on a night</b>,
 * which {@code alarmPlan.js} enforces at the point of editing. If two alarms
 * ever share a day, the window has two answers and this whole approach stops
 * being safe. The JS has the rule and the tests; this is its twin and the two
 * must agree.
 */
final class AlarmPlan {

    /** The whole list, as the app mirrored it. See {@link Entry} for the shape. */
    static final String KEY_PLAN = "alarm_plan";

    /** The band slot tonight's alarm owns, or -1 when nothing is due. */
    static final String KEY_SLOT = "alarm_slot";

    private AlarmPlan() {}

    /**
     * One alarm, as the app mirrors it.
     *
     * <p>Fixed field order, {@code |} separated, alarms separated by {@code ;}:
     * {@code hour|minute|enabled|mode|days|onsetHours|latestHour|latestMinute},
     * where {@code days} is a comma list with Sunday as 0 and an empty string
     * meaning a one-off.
     *
     * <p><b>Flat text rather than JSON, for two reasons.</b> {@code org.json}
     * is an android.jar stub under plain JUnit, so anything parsing it cannot
     * be unit tested without pulling in a dependency; and the same argument
     * {@link AlarmLog} makes applies here, that this is read by
     * {@code run-as cat} as often as by code.
     */
    private static final class Entry {
        int hour = -1;
        int minute;
        boolean enabled;
        String mode = "fixed";
        String days = "";
        float onsetHours = 8f;
        int latestHour = -1;
        int latestMinute;
    }

    /**
     * Choose tonight's alarm and write it into the single-alarm keys.
     *
     * <p>Called before anything reads those keys, so the service always sees
     * the alarm that applies to the day it is currently in. Cheap enough to run
     * every tick, which is what keeps it correct across midnight without
     * anything having to notice midnight happened.
     *
     * @return the slot written, or -1 when the plan is empty or all-off.
     */
    static int resolveActive(final SharedPreferences prefs, final long now) {
        final String raw = prefs.getString(KEY_PLAN, null);
        if (raw == null || raw.isEmpty()) return prefs.getInt(KEY_SLOT, 0);

        final Entry[] list = parse(raw);
        // A plan that parsed to nothing is a malformed one. Leaving the
        // previous keys in place is the safe failure and the whole reason this
        // feature is safe at all: the strap rings the time on its face
        // regardless, so a bad mirror must never be worse than no mirror.
        if (list.length == 0) return prefs.getInt(KEY_SLOT, 0);

        int slot = dueOn(list, now);
        // Nothing today. Look ahead for something to schedule toward - dueToday
        // reads the days written below and correctly answers false for it.
        for (int ahead = 1; ahead <= 7 && slot < 0; ahead++) {
            final Calendar c = Calendar.getInstance();
            c.setTimeInMillis(now);
            c.add(Calendar.DAY_OF_YEAR, ahead);
            slot = dueOn(list, c.getTimeInMillis());
        }
        if (slot < 0) {
            prefs.edit()
                    .putBoolean(SmartAlarm.KEY_ENABLED, false)
                    .putInt(KEY_SLOT, -1)
                    .apply();
            return -1;
        }
        write(prefs, list[slot], slot);
        return slot;
    }

    private static Entry[] parse(final String raw) {
        final String[] parts = raw.split(";");
        final java.util.List<Entry> out = new java.util.ArrayList<>();
        for (final String part : parts) {
            if (part.isEmpty()) continue;
            // Split on a literal pipe: it is a regex alternation otherwise, and
            // an unescaped one matches the empty string between every character.
            final String[] f = part.split("\\|", -1);
            if (f.length < 8) return new Entry[0];
            final Entry e = new Entry();
            try {
                e.hour = Integer.parseInt(f[0]);
                e.minute = Integer.parseInt(f[1]);
                e.enabled = "1".equals(f[2]);
                e.mode = f[3];
                e.days = f[4];
                e.onsetHours = Float.parseFloat(f[5]);
                e.latestHour = Integer.parseInt(f[6]);
                e.latestMinute = Integer.parseInt(f[7]);
            } catch (final NumberFormatException bad) {
                // One unreadable alarm makes the whole plan untrustworthy: the
                // rest may still parse, but which alarm owns which slot would
                // be a guess, and a guess here writes over the wrong slot.
                return new Entry[0];
            }
            out.add(e);
        }
        return out.toArray(new Entry[0]);
    }

    /** Index of the enabled alarm covering that day, or -1. */
    private static int dueOn(final Entry[] list, final long millis) {
        final Calendar c = Calendar.getInstance();
        c.setTimeInMillis(millis);
        final int today = c.get(Calendar.DAY_OF_WEEK) - 1;
        for (int i = 0; i < list.length; i++) {
            final Entry e = list[i];
            if (!e.enabled) continue;
            // An empty list is a one-off, which the band reads as REPEAT_ONCE
            // and which counts as due whatever day it is.
            if (e.days.isEmpty()) return i;
            for (final String d : e.days.split(",")) {
                try {
                    if (Integer.parseInt(d.trim()) == today) return i;
                } catch (final NumberFormatException ignored) {
                    // A malformed day is not a reason to fire on the wrong one.
                }
            }
        }
        return -1;
    }

    /** Write one alarm into the keys {@link SmartAlarm} already reads. */
    private static void write(final SharedPreferences prefs, final Entry e, final int slot) {
        prefs.edit()
                .putString(SmartAlarm.KEY_MODE, e.mode)
                .putInt(SmartAlarm.KEY_HOUR, e.hour)
                .putInt(SmartAlarm.KEY_MINUTE, e.minute)
                .putBoolean(SmartAlarm.KEY_ENABLED, e.enabled)
                .putString(SmartAlarm.KEY_DAYS, e.days)
                .putFloat(SmartAlarm.KEY_ONSET_HOURS, e.onsetHours)
                .putInt(SmartAlarm.KEY_LATEST_HOUR, e.latestHour)
                .putInt(SmartAlarm.KEY_LATEST_MINUTE, e.latestMinute)
                .putInt(KEY_SLOT, slot)
                .apply();
    }

    /**
     * The slot a write should land on.
     *
     * <p>Defaults to 0, which is what every Atlas build before this one used, so
     * a phone that has not yet mirrored a plan keeps writing where its existing
     * alarm already lives.
     */
    static int activeSlot(final SharedPreferences prefs) {
        final int slot = prefs.getInt(KEY_SLOT, 0);
        return slot < 0 ? 0 : slot;
    }
}

File: src/utils/alarmPlan.js
/**
 * More than one alarm, and which of them tonight belongs to.
 *
 * Atlas held exactly one alarm until 2026-08-24. The ask was a weekday alarm
 * and a different weekend one, and the shape it landed on is a list rather than
 * two named schedules, because "weekday" and "weekend" are an arbitrary split
 * and a list is the model every phone already uses.
 *
 * **The band really does have slots.** Proven on the device rather than assumed:
 * a write to slot 1 was accepted, the strap buzzed for it, and Zepp showed both
 * alarms afterwards. So each alarm gets its own slot and the band rings each one
 * itself - nothing has to be rewritten as the week turns.
 *
 * <h2>The one rule everything rests on</h2>
 *
 * **No two enabled alarms may claim the same day.** The smart window, the
 * fired-once mark and the re-arm debt are all single values in the background
 * service, and they can stay that way only while at most one alarm is due on
 * any given night. Allowing two would mean the window has two answers and the
 * service has to keep per-alarm state through a night it is not awake for.
 *
 * That is a real restriction and it is enforced here rather than explained
 * away: a day already spoken for is refused at the point of editing.
 */

/**
 * The cap, and why it is not larger.
 *
 * Slot 1 is proven; how many the firmware actually has is not. Atlas has never
 * read the band's own config, so the honest number is "more than one, unknown
 * how many". Five is comfortably inside any plausible count and there are only
 * seven days to spread across anyway. Raise it once the read side exists.
 */
export const MAX_ALARMS = 5;

/** Sunday 0, matching `Date.getDay()` and what the app mirrors to the band. */
export const DAY_LETTERS = ["S", "M", "T", "W", "T", "F", "S"];
export const DAY_NAMES = ["SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT"];

/** A blank alarm, for the ADD row. Defaults match the old single-alarm ones. */
export function emptyAlarm(id) {
  return {
    id,
    hour: 8,
    minute: 30,
    days: [],
    enabled: true,
    mode: "fixed",
    onsetHours: 8,
    latestHour: null,
    latestMinute: null,
  };
}

/**
 * Which days are claimed by more than one enabled alarm.
 *
 * Returns day numbers, so a caller can both refuse the save and light the
 * offending letters rather than printing a sentence about them.
 */
export function clashingDays(alarms, { ignoreId = null } = {}) {
  const seen = new Map();
  const clashes = new Set();
  for (const a of alarms ?? []) {
    if (!a?.enabled || a.id === ignoreId) continue;
    for (const day of a.days ?? []) {
      if (seen.has(day)) clashes.add(day);
      seen.set(day, a.id);
    }
  }
  return [...clashes].sort((x, y) => x - y);
}

/** The days another enabled alarm already owns, for greying them while editing. */
export function daysTakenBy(alarms, exceptId) {
  const taken = new Set();
  for (const a of alarms ?? []) {
    if (!a?.enabled || a.id === exceptId) continue;
    for (const day of a.days ?? []) taken.add(day);
  }
  return taken;
}

/**
 * Whether a list can be saved, and what is wrong when it cannot.
 *
 * An alarm with no days at all is a one-off, which the band understands and
 * which cannot clash with anything, so it is allowed.
 */
export function validate(alarms) {
  const list = alarms ?? [];
  const errors = [];
  if (list.length > MAX_ALARMS) {
    errors.push(`ATLAS HOLDS ${MAX_ALARMS} ALARMS`);
  }
  const clashes = clashingDays(list);
  if (clashes.length) {
    errors.push(
      `${clashes.map((d) => DAY_NAMES[d]).join(", ")} HAS TWO ALARMS`
    );
  }
  for (const a of list) {
    if (!Number.isInteger(a?.hour) || a.hour < 0 || a.hour > 23) {
      errors.push("AN ALARM HAS NO VALID TIME");
      break;
    }
  }
  return { ok: errors.length === 0, errors, clashes };
}

/**
 * The alarm due on a given date, or null.
 *
 * An empty day list is a one-off and counts as due every day, which matches
 * how the band reads {@code REPEAT_ONCE} and how the service already treats it.
 */
export function alarmDueOn(alarms, date) {
  const day = date.getDay();
  for (const a of alarms ?? []) {
    if (!a?.enabled) continue;
    const days = a.days ?? [];
    if (days.length === 0 || days.includes(day)) return a;
  }
  return null;
}

/**
 * The time an alarm's face shows, which is not always the time it computes.
 *
 * An onset alarm hands the band the *latest* time the user was willing to
 * accept, because the strap is holding a backstop rather than the answer. This
 * is the same split `hardHour`/`hardMinute` make on the Java side and the two
 * must agree, or a re-arm puts a fixed alarm at the onset cap.
 */
export function hardTime(alarm) {
  if (!alarm) return null;
  const useLatest = alarm.mode === "onset" && alarm.latestHour != null;
  return {
    hour: useLatest ? alarm.latestHour : alarm.hour,
    minute: useLatest ? (alarm.latestMinute ?? 0) : alarm.minute,
  };
}

/**
 * The next alarm that will actually go off, and when.
 *
 * Walks forward a week from `now`, which is the only way to answer it when the
 * days are a mask: today's alarm may already have passed, and the next one may
 * be six days out.
 */
export function nextDue(alarms, now = new Date()) {
  for (let ahead = 0; ahead <= 7; ahead++) {
    const day = new Date(now);
    day.setDate(day.getDate() + ahead);
    const alarm = alarmDueOn(alarms, day);
    if (!alarm) continue;
    const time = hardTime(alarm);
    const at = new Date(day);
    at.setHours(time.hour, time.minute, 0, 0);
    if (at.getTime() <= now.getTime()) continue;
    return { alarm, at };
  }
  return null;
}

/**
 * The band slot an alarm owns.
 *
 * The list index, which means reordering the list would rewrite slots. Nothing
 * reorders it: alarms are appended and removed, and a removed one leaves its
 * slot to be reclaimed by the next add. That is deliberate - a slot is a
 * physical thing on the band, and shuffling them would silently write over
 * alarms the user did not touch.
 */
export function slotOf(alarms, id) {
  const at = (alarms ?? []).findIndex((a) => a?.id === id);
  return at < 0 ? null : at;
}

/**
 * The old single alarm as a list.
 *
 * Runs once, on the first read after the update. The old record has no `id`,
 * which is exactly how it is recognised.
 */
export function migrateSingle(old) {
  if (!old) return [];
  return [{ ...emptyAlarm("alarm-1"), ...old, id: old.id ?? "alarm-1" }];
}

/**
 * When the next alarm rings, in words.
 *
 * **The countdown is the useful half.** `NEXT` on its own repeats what the day
 * letters already imply; what nobody can work out at 23:00 on a Sunday is how
 * long they have got.
 *
 * Returns null when nothing will ring, so a caller renders nothing rather than
 * an empty row.
 */
export function describeNext(alarms, now = new Date()) {
  const due = nextDue(alarms, now);
  if (!due) return null;

  const midnight = new Date(now);
  midnight.setHours(0, 0, 0, 0);
  const dayMidnight = new Date(due.at);
  dayMidnight.setHours(0, 0, 0, 0);
  const daysAhead = Math.round((dayMidnight - midnight) / 86_400_000);

  let when;
  if (daysAhead === 0) when = "TODAY";
  else if (daysAhead === 1) when = "TOMORROW";
  else when = DAY_NAMES[due.at.getDay()];

  const totalMinutes = Math.max(0, Math.round((due.at - now) / 60_000));
  const hours = Math.floor(totalMinutes / 60);
  const mins = totalMinutes % 60;
  // Minutes alone under an hour, because "0H 45M" is a worse way of saying 45
  // minutes. Hours alone once it is far enough out that the minutes are noise.
  let gap;
  if (hours === 0) gap = `${mins}M`;
  else if (hours >= 24) gap = `${Math.round(hours / 24)}D`;
  else if (mins === 0) gap = `${hours}H`;
  else gap = `${hours}H ${mins}M`;

  return { alarm: due.alarm, at: due.at, when, gap, text: `${when} · IN ${gap}` };
}

/**
 * Which alarm owns a given day, for naming it rather than saying "another one".
 *
 * "Greyed days belong to another alarm" raises the question it does not answer.
 * This is what lets the panel say `MON-FRI BELONG TO YOUR 07:00 ALARM`.
 */
export function ownerOfDays(alarms, exceptId) {
  const owners = new Map();
  for (const a of alarms ?? []) {
    if (!a?.enabled || a.id === exceptId) continue;
    for (const day of a.days ?? []) owners.set(day, a);
  }
  return owners;
}

/**
 * A run of day numbers as people say it: `MON-FRI`, `SAT SUN`, `WED`.
 *
 * Monday-first, because a week reads that way even though the numbers are
 * Sunday-first. Contiguous runs of three or more collapse to a dash; anything
 * else is listed, since `SUN-MON` for a weekend is worse than naming both.
 */
export function describeDayRun(days) {
  const order = [1, 2, 3, 4, 5, 6, 0];
  const present = order.filter((d) => days.includes(d));
  if (!present.length) return "";
  const runs = [];
  let run = [present[0]];
  for (let i = 1; i < present.length; i++) {
    const prevAt = order.indexOf(present[i - 1]);
    if (order.indexOf(present[i]) === prevAt + 1) run.push(present[i]);
    else {
      runs.push(run);
      run = [present[i]];
    }
  }
  runs.push(run);
  return runs
    .map((r) =>
      r.length >= 3
        ? `${DAY_NAMES[r[0]]}-${DAY_NAMES[r[r.length - 1]]}`
        : r.map((d) => DAY_NAMES[d]).join(" ")
    )
    .join(" ");
}

/**
 * The list as the background service reads it.
 *
 * Fixed field order, `|` separated, alarms separated by `;`:
 * `hour|minute|enabled|mode|days|onsetHours|latestHour|latestMinute`.
 *
 * **Flat text rather than JSON, and `AlarmPlan.java` is the twin that parses
 * it.** `org.json` is an android.jar stub under plain JUnit, so a JSON plan
 * could not be unit tested on the Java side without a new dependency - and the
 * same argument `AlarmLog` makes applies, that this gets read by `run-as cat`
 * as often as by code. **Change this and change the Java parser.**
 *
 * Order is the slot order, so this must serialise the list as it stands rather
 * than sorting it: a slot is a physical thing on the band.
 */
export function serialisePlan(alarms) {
  return (alarms ?? [])
    .map((a) =>
      [
        a.hour ?? -1,
        a.minute ?? 0,
        a.enabled ? 1 : 0,
        a.mode ?? "fixed",
        (a.days ?? []).join(","),
        Number(a.onsetHours ?? 8).toFixed(1),
        a.latestHour ?? -1,
        a.latestMinute ?? 0,
      ].join("|")
    )
    .join(";");
}

/** What the closed settings row says. */
export function planSummary(alarms) {
  const on = (alarms ?? []).filter((a) => a?.enabled);
  if (!on.length) return "NONE SET";
  const times = on
    .map((a) => {
      const t = hardTime(a);
      return `${String(t.hour).padStart(2, "0")}:${String(t.minute).padStart(2, "0")}`;
    })
    .sort();
  return times.join(" · ");
}
